src/main/java/entities/Matcher.java

```diff
@@ -0,0 +1,104 @@
+package entities;
+
+import data_access_storage.UserRepo;
+import data_access_storage.UserRepoManager;
+import email_request.*;
+import matching_system.*;
+
+import javax.swing.*;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public class Matcher {
+//    private HashMap<User, Integer> matchResults = new HashMap<User, Integer>();
+    /** The matched result, arranged by matching score from high to low */
+    private final UserAccount[] matches;
+
+    public Matcher(UserAccount matchee, ArrayList<UserAccount> targets){
+        this.matches = match(matchee,targets);
+    }
+
+    public UserAccount[] getMatches() {
+        return matches;
+    }
+
+    /** get matched results in a list
+      @return ArrayList that stores matched results, sorted by score from high to low
+     */
+    private UserAccount[] match(UserAccount matchee, ArrayList<UserAccount> targets){
+        HashMap<UserAccount, Integer> matchResults = new HashMap<UserAccount, Integer>();
+        User matchUser = matchee.getUser();
+        for(UserAccount target :targets){
+            User targetUser = target.getUser();
+            int score = 0;
+            if(targetUser.getGender().equals(matchUser.getInterestedIn())&&
+                    matchUser.getGender().equals(targetUser.getInterestedIn())){
+                score +=1;
+                if(targetUser.getHobby().equals(matchUser.getHobby())){
+                    score +=1;
+                }
+                if(targetUser.getProgramOfStudy().equals(matchUser.getProgramOfStudy())){
+                    score+=1;
+                }
+            }
```

Comment:
At the beginning we intended to have a "preference" for each user, the preference includes user's perfered age. As we decided not to include the preference part, matching age seems weird as simply matching user with others who has the same age might not be what the user want. Also, in terms of the code smells part, I will make this part a seperate method. Is there another way to do this so that I can remove all these if statements? It will be great if it is possible to not use all these if statements.

Reply:
> I think we also mentioned that the users are matched with the same age, right?

At the beginning we intended to have a "preference" for each user, the preference includes user's perfered age. As we decided not to include the preference part, matching age seems weird as simply matching user with others who has the same age might not be what the user want. Also, in terms of the code smells part, I will make this part a seperate method. Is there another way to do this so that I can remove all these if statements? It will be great if it is possible to not use all these if statements.

Reply:
yes, there is a way to remove the if conditions. But the number of if conditions here are not large enough to make those extra efforts. In case there are many if conditions, we can consider extracting classes from the if conditions. I think state design pattern can be a potential solution. look at  https://refactoring.guru/design-patterns/state